model/src/main/scala/ch.epfl.scala.index.model/release/BackendVersion.scala

```diff
@@ -0,0 +1,98 @@
+package ch.epfl.scala.index.model
+package release
+
+import fastparse.NoWhitespace._
+import fastparse._
+
+sealed trait BackendVersion {
+  // dotty versions are indexed under the 'dotty' family
+  def family: String
+  def render: String
+  def isValid: Boolean
+}
+
+final case class ScalaBackend(version: BinaryVersion) extends BackendVersion {
+  def family: String = toString()
+  def render: String = s"scala $version"
+  def isValid: Boolean = ScalaBackend.isValid(version)
+  override def toString: String = version.toString()
+}
+
+final case class DottyBackend(version: BinaryVersion) extends BackendVersion {
+  def family = "dotty"
+  def render: String = s"dotty $version"
+  def isValid: Boolean = DottyBackend.isValid(version)
+  override def toString: String = version.toString()
+}
+
+object ScalaBackend {
+  val `2.10`: ScalaBackend = ScalaBackend(MinorBinary(2, 10))
+  val `2.11`: ScalaBackend = ScalaBackend(MinorBinary(2, 11))
+  val `2.12`: ScalaBackend = ScalaBackend(MinorBinary(2, 12))
+  val `2.13`: ScalaBackend = ScalaBackend(MinorBinary(2, 13))
+
+  private val stableBinaryVersions =
+    Set(`2.10`.version, `2.11`.version, `2.12`.version, `2.13`.version)
+
+  def isValid(version: BinaryVersion): Boolean =
+    stableBinaryVersions.contains(version)
+
+  def Parser[_: P]: P[ScalaBackend] = {
+    BinaryVersion.Parser
+      .filter(isValid)
+      .map(ScalaBackend(_))
+  }
+}
+
+object DottyBackend {
+  val `0.16`: DottyBackend = DottyBackend(MinorBinary(0, 16))
+  val `0.17`: DottyBackend = DottyBackend(MinorBinary(0, 17))
+  val `0.18`: DottyBackend = DottyBackend(MinorBinary(0, 18))
+  val `0.19`: DottyBackend = DottyBackend(MinorBinary(0, 19))
+  val `0.20`: DottyBackend = DottyBackend(MinorBinary(0, 20))
+  val `0.21`: DottyBackend = DottyBackend(MinorBinary(0, 21))
+  val `0.22`: DottyBackend = DottyBackend(MinorBinary(0, 22))
+  val `0.23`: DottyBackend = DottyBackend(MinorBinary(0, 23))
+
+  private val latestVersions =
+    Set(
+      `0.16`.version,
+      `0.17`.version,
+      `0.18`.version,
+      `0.19`.version,
+      `0.20`.version,
+      `0.21`.version,
+      `0.22`.version,
+      `0.23`.version
+    )
+
+  def isValid(version: BinaryVersion): Boolean =
+    latestVersions.contains(version) ||
+      (version > `0.23`.version && version < MinorBinary(0, 30))
```

Comment:
Why not just this?

```suggestion
  def isValid(version: BinaryVersion): Boolean = version > `0.16`.version
```

Reply:
The reasons are:
- we want to distiguinsh between scala version and dotty version: `2.12` is not a valid dotty version
- previous milestones and RC are not relevant: `0.22-RC1` is not a valid dotty version
but a future `0.24-RC1` is a valid dotty version
- some sbt plugin have bad version: `_0.31_2.12` instead of `_2.12_0.31` and we don't want to interpret this as a plugin for sbt 2.12 and dotty 0.31

Reply:
I see, thanks for the explanation!

Does that mean that we will have to update scaladex each time a new Dotty RC is released?

Reply:
That's a good question. As of now I have seen no libraries issued with a Dotty RC version. But we will have to update Scaladex when Dotty version 0.30 will be released. And this is also true for scala 3.0 scala.js 2.x and scala-native 0.4 (not sure for this one) and sbt 2.x.

Hopefully we will make the ScalaTarget parser more robust and then we will make it forward compatible.

Reply:
Maybe we can assume for now that Dotty libraries will always target non-RC Dotty releases? So maybe we can write a more general test that checks that the major version is 0 and that the patch version is a number (it doesn’t contain “RC...”). Wdyt?

Reply:
Yes it would be better. But we still want to discard the versions older than 0.17 and the badly formated sbt versions.

Reply:
For badly formatted sbt versions, I believe we could make a more robust parser: it should only accept one binary suffix, not two.

Reply:
I don't really know where do those sbt target versions come frome but they have two parts, a scala version and a sbt version. Most of the time we have _<scala version>_<sbt version> but sometimes they are inverted.

Reply:
The thing that I don’t understand is why is a pattern such as `_x.y_α.β` successfully parsed in the first place as a valid Scala version?

Reply:
Its not a Scala version its a Sbt target version. Nevertheless I've moved the logic of discarding bad sbt plugin target to SbtPlugin which is indeed better.